raiden/connection_manager.py

```diff
@@ -265,39 +269,84 @@ def _open_channels(self):
+            ]
+            nonfunded_channels = [
+                channel_state for channel_state in open_channels
+                if channel_state not in funded_channels
+            ]
+            possible_new_partners = self.find_new_partners()
+
+            # if we already met our target, break
+            if len(funded_channels) >= self.initial_channel_target:
+                break
+            # if we didn't, but there's no nonfunded channels and no available partners
+            # it means the network is smaller than our target, so we should also break
+            if not nonfunded_channels and not possible_new_partners:
+                break
+
+            n_to_join = self.initial_channel_target - len(funded_channels)
+            nonfunded_partners = [
+                channel_state.partner_state.address
+                for channel_state in nonfunded_channels
+            ]
+            # first, fund nonfunded channels, then open and fund with possible_new_partners,
+            # until initial_channel_target of funded channels is met
+            join_partners = (nonfunded_partners + possible_new_partners)[:n_to_join]
+
+            def _join_partner(partner):
+                has_channel = any(
+                    channel_state.partner_state.address == partner
+                    for channel_state in open_channels
                 )
-            except AddressWithoutCode:
-                log.warn('connection manager: channel closed just after it was created')
-            except TransactionThrew:
-                log.exception('connection manager: deposit failed')
+                if not has_channel:
+                    try:
+                        self.api.channel_open(
+                            self.registry_address,
+                            self.token_address,
+                            partner,
+                        )
+                    except DuplicatedChannelError:
+                        # This can fail because of a race condition, where the channel
+                        # partner opens first.
+                        log.info('partner opened channel first')
+
+                try:
+                    self.api.set_total_channel_deposit(
+                        self.registry_address,
+                        self.token_address,
+                        partner,
+                        self._initial_funding_per_partner,
+                    )
+                except AddressWithoutCode:
+                    log.warn('connection manager: channel closed just after it was created')
+                except TransactionThrew:
+                    log.exception('connection manager: deposit failed')
+                except ChannelIncorrectStateError:
+                    log.exception('connection manager: channel not in opened state')
+                else:
+                    return True
+
+            greenlets = [
```

Comment:
what if the gevent fails? will the exception propagate to the central hub? 
How about checking return value of `_join_partner()` and logging at least a warning if something went wrong?

Reply:
They will propagate up the stack. These exceptions handled there are the only ones that may happen sometimes, and are harmless, as it's run in a loop.